dasharo-stability/capsule-update.robot

```diff
@@ -89,11 +103,54 @@ CUP999.001 Capsule Update
     Should Contain    ${out}    CapsuleMax
     Should Not Contain    ${out}    CapsuleLast
 
+CUP160.001 Verifying UUID
+    [Documentation]    Check if UUID didn't change after Capsule Update.
+    Go To Ubuntu Prompt
```

Comment:
This is common to all of post-update tests except for BIOS settings persistence. This keyword takes 2:45 each time, which means additional 5:30 to the test suite execution time.

My proposition:
- make `Get System Values` take arguments which would specify the name of suite variables to write to (remember to **not** use escaped versions, as not recommended in https://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Using%20variables%20with%20keywords%20creating%20or%20accessing%20variables)
- use `Get System Values    '$ORIGINAL_UUID'    '$ORIGINAL_SERIAL'    '$CUSTOM_LOGO_SHA'` in suite setup (I'm not sure about how exactly those names should be escaped)
- in these 3 tests, change first two lines (more lines in case of logo) to something like:
    ```
    Run Keyword If    $UPDATED_UUID is None    Get System Values    '$UPDATED_UUID'    '$UPDATED_SERIAL'    '$UPDATED_LOGO_SHA'
    ```
    Of course, change `$UPDATED_UUID is None` to whatever current test checks.

This should make it boot only once, while still allowing to skip some tests when not needed.

Reply:
> lambda function (like in C)

C doesn't have lambda functions.

What I'm asking is to add arguments that are **names** of variables, that would be expanded as per linked documentation. If it is mentioned in the official documentation, chances are that RF has such capability :slightly_smiling_face: 

> Or I could merge those 3 tests to 1.

No, that way we wouldn't know which one failed without looking in the logs. Or worse, logic for skipping a part of the test would skip something that should be there.

Reply:
@JanPrusinowski I've tested the variable handling and discovered that:

- in the keyword invocation, variable names must be passed without any quotation and curly braces (e.g. `$ORIGINAL_UUID`), no other form worked for me
- in the `[Arguments]`, I used `${var_uuid}` form, haven't checked other forms but this one works
- in the body of keyword, `${var_uuid}` form must be used when calling `Set Suite Variable` to have the name properly expanded.

I've updated my comment above accordingly, let me know if you still have any problems with it.

Reply:
I have made changes to the keyword in: 2e122f2df5630f9b82de24c8073696d619eb36cf
Let me know it it is what you are looking for. Anyway I think that this overcomplicates things...

Reply:
Starting from https://github.com/Dasharo/open-source-firmware-validation/commit/e7ab746fb5c8ed7ab99ba02fd985844182bd588e:

```diff
diff --git a/dasharo-stability/capsule-update.robot b/dasharo-stability/capsule-update.robot
index fca5bbe51e72..fd15402650a2 100644
--- a/dasharo-stability/capsule-update.robot
+++ b/dasharo-stability/capsule-update.robot
@@ -33,7 +33,7 @@ Suite Setup         Run Keywords
 ...                     AND
 ...                     Upload Required Files
 ...                     AND
-...                     Get System Values
+...                     Get System Values    $ORIGINAL_SERIAL    $ORIGINAL_UUID    $ORIGINAL_LOGO_SHA256
 ...                     AND
 ...                     Turn Off Active ME
 Suite Teardown      Run Keywords
@@ -115,38 +115,31 @@ CUP160.001 Verifying BIOS Settings Persistence After Update - PART 2
 
 CUP170.001 Verifying UUID
     [Documentation]    Check if UUID didn't change after Capsule Update.
-    Go To Ubuntu Prompt
+    Run Keyword If    $UPDATED_UUID is None    Get System Values    $UPDATED_UUID    $UPDATED_SERIAL    $UPDATED_LOGO_SHA256
+
     Log To Console    \n[Before Update] ${ORIGINAL_UUID}
-    ${updated_uuid}=    Get Firmware UUID
-    Log To Console    \n[After Update] ${updated_uuid}
+    Log To Console    \n[After Update] ${UPDATED_UUID}
 
-    Should Be Equal    ${ORIGINAL_UUID}    ${updated_uuid}
+    Should Be Equal    ${ORIGINAL_UUID}    ${UPDATED_UUID}
     IF    ${ROMHOLE_SUPPORT} == ${TRUE}
-        Should Be Equal    ${ORIGINAL_UUID}    00112233-4455-6677-8899-aabbccddeeff
-        Should Be Equal    ${updated_uuid}    00112233-4455-6677-8899-aabbccddeeff
+        Should Be Equal    ${UPDATED_UUID}    00112233-4455-6677-8899-aabbccddeeff
     END
 
 CUP180.001 Verifying Serial Number
     [Documentation]    Check if serial number didn't change after Capsule Update.
-    Go To Ubuntu Prompt
+    Run Keyword If    $UPDATED_SERIAL is None    Get System Values    $UPDATED_UUID    $UPDATED_SERIAL    $UPDATED_LOGO_SHA256
+
     Log To Console    \n[Before Update] ${ORIGINAL_SERIAL}
-    ${updated_serial}=    Get Firmware Serial Number
-    Log To Console    \n[After Update] ${updated_serial}
+    Log To Console    \n[After Update] ${UPDATED_SERIAL}
 
-    Should Be Equal    ${ORIGINAL_SERIAL}    ${updated_serial}
+    Should Be Equal    ${ORIGINAL_SERIAL}    ${UPDATED_SERIAL}
 
 CUP190.001 Verifying If Custom Logo Persists Across updates
     [Documentation]    Check if Logo didn't change after Capsule Update.
     Skip If    not ${CUSTOM_LOGO_SUPPORT}    not supported
-    Go To Ubuntu Prompt
-    ${out}=    Execute Command In Terminal
-    ...    sha256sum /sys/firmware/acpi/bgrt/image
-    ${unplugged}=    Run Keyword And Return Status
-    ...    Should Contain    ${out}    No such file
-    IF    ${unplugged} == ${TRUE}
-        Fail    Please make sure that a display device is connected to the DUT
-    END
-    Should Contain    ${out}    ${CUSTOM_LOGO_SHA}
+    Run Keyword If    $UPDATED_LOGO_SHA256 is None    Get System Values    $UPDATED_UUID    $UPDATED_SERIAL    $UPDATED_LOGO_SHA256
+
+    Should Be Equal    ${ORIGINAL_LOGO_SHA256}    ${UPDATED_LOGO_SHA256}
 
 
 *** Keywords ***
@@ -380,15 +373,11 @@ Go To Ubuntu Prompt
     Switch To Root User
 
 Get System Values
-    Log To Console    PREPARE: Get System Values
+    [Arguments]    ${var_serial}    ${var_uuid}    ${var_logo_sha256}
 
-    ${temp_serial}=    Get Firmware Serial Number
-    Set Suite Variable    ${ORIGINAL_SERIAL}    ${temp_serial}
-    Log To Console    \n[Serial Number Before Update] ${ORIGINAL_SERIAL}
+    Set Suite Variable    ${var_serial}    Get Firmware Serial Number
 
-    ${temp_uuid}=    Get Firmware UUID
-    Set Suite Variable    ${ORIGINAL_UUID}    ${temp_uuid}
-    Log To Console    [UUID Before Update] ${ORIGINAL_UUID}\n
+    Set Suite Variable    ${var_uuid}    Get Firmware UUID
 
     IF    $CUSTOM_LOGO_SUPPORT == $TRUE
         ${out}=    Execute Command In Terminal
@@ -398,7 +387,7 @@ Get System Values
         IF    ${unplugged} == ${TRUE}
             Fail    Please make sure that a display device is connected to the DUT
         END
-        Set Suite Variable    ${CUSTOM_LOGO_SHA}    ${out}
+        Set Suite Variable    ${var_logo_sha256}    ${out}
     END
 
 Prepare For ROMHOLE Persistence Test

```

Reply:
Oh, now `Get System Values` doesn't boot to Linux, this would have to be added. Instead of all three `Run Keyword If` you would have to do something like:

```python
IF    $UPDATED_UUID is None
    Go To Ubuntu Prompt
    Get System Values    $UPDATED_UUID    $UPDATED_SERIAL    $UPDATED_LOGO_SHA256
END
```

Reply:
Updated code in https://github.com/Dasharo/open-source-firmware-validation/pull/501#discussion_r1778344724, `$UPDATED_UUID` isn't `None` without `Get Variable Value`. I also messed up the order of arguments, this was also fixed in that comment.

Reply:
I see serial number, but remember that custom romhole contents doesn't set it.  It's part of this structure which isn't trivial to construct and looks like it should be board-specific, so I left it filled with `0xff`.  At least for 
```c
struct smbios_type2_v3_4 {
	struct smbios_header header;
	u8 manufacturer;
	u8 product_name;
	u8 version;
	u8 serial_number;
	u8 asset_tag;
	u8 feature_flags;
	u8 location_in_chassis;
	u16 chassis_handle;
	u8 board_type;
	u8 contained_objects;
	u8 eos[2];
} __packed;
```
So for MSI serial number check should be skipped or we need to make this struct with somewhat valid data.

Reply:
I have made some changes acording to @krystian-hebel suggestions.
Also noticed that there is no boot to ubuntu - but I have changed the order of functions in prepare and the keyword should work now.
6a465ff4c5e22311bb4b39ea6c07d9fb1c1b8b67

Reply:
@JanPrusinowski this is from previous version, I updated the comment later. UUID and serial have switched places.

Reply:
Also removed not needed resource file: 695a31bbc62ddc420a10a59b526859277c943535